components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java

```diff
@@ -0,0 +1,374 @@
+        } catch (Exception e) {
+            throw new IdentityKeyStoreResolverException(
+                    "Error occurred when retrieving keystore for tenant: " + tenantDomain, e);
+        }
+    }
+
+    /**
+     * Return Primary, tenant or custom keystore.
+     *
+     * @param tenantDomain      Tenant domain
+     * @param inboundProtocol   Inbound authentication protocol of the application
+     * @return Primary, tenant or custom keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public KeyStore getKeyStore(String tenantDomain, InboundProtocol inboundProtocol)
+            throws IdentityKeyStoreResolverException {
+
+        if (keyStoreMappings.containsKey(inboundProtocol)) {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain) ||
+                    keyStoreMappings.get(inboundProtocol).getUseInAllTenants()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Custom keystore configuration avialble for " + inboundProtocol + " protocol.");
+                }
+
+                String keyStoreName = buildCustomKeyStoreName(keyStoreMappings.get(inboundProtocol).getKeyStoreName());
+
+                try {
+                    int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+                    KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+                    return keyStoreManager.getKeyStore(keyStoreName);
+                } catch (Exception e) {
+                    throw new IdentityKeyStoreResolverException(
+                            "Error occurred when retrieving keystore for protocol: " + inboundProtocol, e);
+                }
+            }
+        }
+
+        return getKeyStore(tenantDomain);
+    }
+
+    /**
+     * Return Primary key of the Primary or tenant keystore according to given tenant domain.
+     *
+     * @param tenantDomain  Tenant domain
+     * @return Primary key of Primary or tenant keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public Key getPrivateKey(String tenantDomain) throws IdentityKeyStoreResolverException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        if (privateKeys.containsKey(String.valueOf(tenantId))) {
+            return privateKeys.get(String.valueOf(tenantId));
+        }
+
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        Key privateKey;
+
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                privateKey = keyStoreManager.getDefaultPrivateKey();
+            } else {
+                String tenantKeyStoreName = buildTenantKeyStoreName(tenantDomain);
+                privateKey = keyStoreManager.getPrivateKey(tenantKeyStoreName, tenantDomain);
+            }
+        } catch (Exception e) {
```

Comment:
Catch specific exceptions not generic ones "Exception"

Reply:
`KeyStoremanager.getPrivateKey` throws the common `Exception` type.